app/controllers/application_controller.rb

```diff
@@ -1,10 +1,147 @@
 require './config/environment'
-
+require "./app/models/user"  #not sure if needed
+require 'pry'
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+
+  get '/' do
+    @user = current_user if is_logged_in?
+    erb :index
+  end
+
+  get '/signup' do
+    if is_logged_in?
+      redirect '/tweets'
+    else
+      erb :'/users/signup'
+    end
+  end
+
+  post '/signup' do
+    user = User.new(:username => params[:username], :email => params[:email], :password => params[:password])
+    if user.save && user.username != "" && user.email != ""
+      session[:user_id] = user.id
+      redirect to "/tweets"
+    else
+      redirect '/signup'
+    end
+    redirect to "/tweets"
+  end
+
+  get '/login' do
+    if is_logged_in?
+      redirect '/tweets'
+    else
+      erb :'/users/login'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(:username => params[:username])
+
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+    end
+    redirect '/tweets'
+  end
+
+  get "/users/:slug" do
+    @user = User.find_by_slug(params[:slug])
+    erb :'/users/show'
+  end
+
+  get '/logout' do
+    session.clear
+    redirect "/login"
+  end
+
+
+  post '/tweet' do
+    if !params[:content].empty?
+      tweet = Tweet.create(:content => params[:content])
+      current_user.tweets << tweet
+      current_user.save
+      redirect '/tweets'
+    else
+      redirect to '/tweets/new'
+    end
+  end
+
+  get '/tweets' do
+    if is_logged_in?
+      @user = current_user
+      @tweets = Tweet.all
+      erb :'/tweets/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/new' do
+    if is_logged_in?
+      @user = current_user
+      erb :'/tweets/create_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/:id' do
+    if is_logged_in?
+      @user = current_user
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'/tweets/show_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    @tweet = Tweet.find_by_id(params[:id])
+    if is_logged_in? && @tweet.user == current_user
+      erb :'/tweets/edit_tweet'
+    else
+      redirect 'login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    @tweet = Tweet.find_by_id(params[:id])
+    if !params[:content].empty?
+      @tweet.update(:content => params[:content])
+      @tweet.save
+      redirect "tweets/#{params[:id]}"
+    else
+      redirect "tweets/#{params[:id]}/edit"
+    end
+  end
+
+  post '/tweets/:id/delete' do
+    @tweet = Tweet.find_by_id(params[:id])
+    if current_user == @tweet.user
+      @tweet.delete
+      redirect to '/tweets'
+    else
+      redirect to "/tweets/#{params[:id]}"
+    end
+  end
+
+
+
+  helpers do
+    def current_user
+      User.find(session[:user_id])
```

Comment:
Because in the lifespan of a request, you might call `current_user` multiple times, we'd prefer this method to only hit the database once and then 'cache' the user instance into an instance variable that could be returned on any subsequent call to `current_user`.

```ruby
def current_user
  @current_user ||= User.find(session[:id]) if session[:id]
end
```

This pattern of conditionally setting an instance variable or returning it if it's already defined using the `||=` (or-equals) operator is called Memoization ([read more](http://gavinmiller.io/2013/basics-of-ruby-memoization/) and [here](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/)).